### PR TITLE
planner: unfold the wildcard when creating view

### PR DESCRIPTION
**Pull request by AndrewDi**
### What problem does this PR solve? 
As part of View implement, remove view's wildcard and expand it.

### What is changed and how it works?
ref proposal [Proposal: Implement View](https://github.com/pingcap/tidb/blob/master/docs/design/2018-10-24-view-support.md)

### Check List 

Tests 

 - Unit test

### Description
Let's take a example with create view,
```mysql
CREATE VIEW V AS SELECT *FROM mysql.tidb
```
Before this PR，View will store select statement like 
```mysql
SELECT * FROM mysql.tidb
```

After this PR, View will store select statement like 
```mysql
SELECT VARIABLE_NAME,VARIABLE_VALUE,COMMENT FROM mysql.tidb
```

### CODE REVIEW

**Comment by AndrewDi** [comment]
@XuHuaiyu PTAL

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -2397,6 +2395,31 @@ func (b *PlanBuilder) BuildDataSourceFromView(ctx context.Context, dbName model.
 		b.visitInfo = appendVisitInfo(b.visitInfo, mysql.ShowViewPriv, dbName.L, tableInfo.Name.L, "", ErrViewNoExplain)
 	}
 
+	if tableInfo.View.Cols == nil {
```
Comment (XuHuaiyu) [inline planner/core/logical_plan_builder.go]: Can we extract the two similar code block to a function?

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -2130,12 +2130,14 @@ func (b *PlanBuilder) buildDDL(ctx context.Context, node ast.DDLNode) (Plan, err
 			return nil, err
 		}
 		schema := plan.Schema()
-		if v.Cols != nil && len(v.Cols) != schema.Len() {
-			return nil, ddl.ErrViewWrongList
+		if v.Cols == nil {
```
Comment (XuHuaiyu) [inline planner/core/planbuilder.go]: If we run `create view v(a,b,c) as selet b,d,e from t`.
v.Cols here is [a,b,c] ?

Comment (AndrewDi) [reply planner/core/planbuilder.go]: yes, we just need to colname to build projection.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -1503,33 +1514,17 @@ func (d *ddl) CreateView(ctx sessionctx.Context, s *ast.CreateViewStmt) (err err
...
-			tableColumns[i] = table.ToColumn(&model.ColumnInfo{
-				Name:    v,
-				ID:      int64(i),
-				Offset:  i,
-				State:   model.StatePublic,
-				Version: model.CurrLatestColumnInfoVersion,
-			})
-		}
+func buildViewInfo(ctx sessionctx.Context, s *ast.CreateViewStmt) (*model.ViewInfo, error) {
+	restoreFlag := format.RestoreStringSingleQuotes | format.RestoreKeyWordUppercase | format.RestoreNameBackQuotes
```
[the remark this answers is not in the document]
Comment (AndrewDi) [reply ddl/ddl_api.go]: ok

File: planner/core/logical_plan_builder.go (hunk 1 above)
Comment (AndrewDi) [reply planner/core/logical_plan_builder.go]: It's not easy, how about this way?

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -1503,33 +1514,18 @@ func (d *ddl) CreateView(ctx sessionctx.Context, s *ast.CreateViewStmt) (err err
...
+func buildViewInfo(ctx sessionctx.Context, s *ast.CreateViewStmt) (*model.ViewInfo, error) {
+	// Always Use `format.RestoreNameBackQuotes` to restore `SELECT` statement despite the `ANSI_QUOTES` SQL Mode is enabled or not.
+	restoreFlag := format.RestoreStringSingleQuotes | format.RestoreKeyWordUppercase | format.RestoreNameBackQuotes
+	var sb strings.Builder
+	if err := s.Select.Restore(format.NewRestoreCtx(restoreFlag, &sb)); err != nil {
+		return nil, err
 	}
 
-	return viewInfo, tableColumns
+	viewInfo := &model.ViewInfo{Definer: s.Definer, Algorithm: s.Algorithm,
```
Comment (zz-jason) [inline ddl/ddl_api.go]: how about:
```go
return &model.ViewInfo{
    Algorithm:   s.Algorithm,
    Definer:     s.Definer,
    Security:    s.Security,
    SelectStmt:  sb.String(),
    CheckOption: s.CheckOption,
}, nil
```

File: planner/core/logical_plan_builder.go [hunk 5]
```diff
@@ -2397,23 +2395,44 @@ func (b *PlanBuilder) BuildDataSourceFromView(ctx context.Context, dbName model.
...
 	}
 
-	projSchema := expression.NewSchema(make([]*expression.Column, 0, len(tableInfo.View.Cols))...)
-	projExprs := make([]expression.Expression, 0, len(tableInfo.View.Cols))
-	for i := range tableInfo.View.Cols {
-		col := selectLogicalPlan.Schema().FindColumnByName(tableInfo.View.Cols[i].L)
-		if col == nil {
-			return nil, ErrViewInvalid.GenWithStackByArgs(dbName.O, tableInfo.Name.O)
+	if len(tableInfo.Columns) != selectLogicalPlan.Schema().Len() {
+		return nil, ErrViewInvalid.GenWithStackByArgs(dbName.O, tableInfo.Name.O)
```
Comment (zz-jason) [inline planner/core/logical_plan_builder.go]: could you add a test for this invalid state?

Comment (AndrewDi) [reply planner/core/logical_plan_builder.go]: This check only valid when wildcard still exists in view definetion.

File: planner/core/logical_plan_builder.go (hunk 1 above)
Comment (XuHuaiyu) [reply planner/core/logical_plan_builder.go]: We can extract 2042~2434 to a function like buildProjUponView.

Comment (XuHuaiyu) [reply planner/core/logical_plan_builder.go]: Like this:
``` go
func (b *PlanBuilder) BuildProjUponView(ctx context.Context, dbName model.CIStr, tableInfo *model.TableInfo, selectLogicalPlan Plan) (Plan, error){
	columnInfo := tableInfo.Cols()
	cols := selectLogicalPlan.Schema().Columns
	// In the old version of VIEW implementation, tableInfo.View.Cols is used to
	// store the origin columns' names of the underlying SelectStmt used when
	// creating the view.
	if tableInfo.View.Cols != nil {
		cols = cols[:0]
		for _, info := range columnInfo {
			col := selectLogicalPlan.Schema().FindColumnByName(info.Name.L)
			if col == nil {
				return nil, ErrViewInvalid.GenWithStackByArgs(dbName.O, tableInfo.Name.O)
			}
			cols = append(cols, col)
		}
	}
	projSchema := expression.NewSchema(make([]*expression.Column, 0, len(tableInfo.Columns))...)
	projExprs := make([]expression.Expression, 0, len(tableInfo.Columns))
	for i, col := range cols {
		origColName := col.ColName
		if tableInfo.View.Cols != nil {
			origColName = tableInfo.View.Cols[i]
		}
		projSchema.Append(&expression.Column{
			UniqueID:    b.ctx.GetSessionVars().AllocPlanColumnID(),
			TblName:     col.TblName,
			OrigTblName: col.OrigTblName,
			ColName:     columnInfo[i].Name,
			OrigColName: origColName,
			DBName:      col.DBName,
			RetType:     col.GetType(),
		})
		projExprs = append(projExprs, col)
	} 
	projUponView := LogicalProjection{Exprs: projExprs}.Init(b.ctx)
	projUponView.SetChildren(selectLogicalPlan.(LogicalPlan))
	projUponView.SetSchema(projSchema)
	return projUponView, nil
}

```

Comment (AndrewDi) [reply planner/core/logical_plan_builder.go]: Thanks

Comment (AndrewDi) [reply planner/core/logical_plan_builder.go]: @XuHuaiyu PTAL again

File: planner/core/logical_plan_builder.go [hunk 6]
```diff
@@ -2176,7 +2175,6 @@ func (b *PlanBuilder) buildSelect(ctx context.Context, sel *ast.SelectStmt) (p L
 		}
 	}
 
-	sel.Fields.Fields = originalFields
```
Comment (XuHuaiyu) [inline planner/core/logical_plan_builder.go]: Is this acceptable? @winoros

File: executor/show_test.go [hunk 7]
```diff
@@ -446,14 +446,14 @@ func (s *testSuite2) TestShowCreateTable(c *C) {
 	tk.MustExec("create table t1(a int,b int)")
 	tk.MustExec("drop view if exists v1")
 	tk.MustExec("create or replace definer=`root`@`127.0.0.1` view v1 as select * from t1")
```
Comment (XuHuaiyu) [inline executor/show_test.go]: please add some test cases like 
``` sql
select avg(a), * from t1;
select a+b, * , a from t1;
```

Comment (AndrewDi) [reply executor/show_test.go]: @XuHuaiyu These test should add to `TestCreateView` in `ddl_test.go`?

Comment (XuHuaiyu) [reply executor/show_test.go]: yes

Comment (AndrewDi) [reply executor/show_test.go]: done

File: ddl/ddl_api.go [hunk 8]
```diff
@@ -1461,11 +1461,22 @@ func (d *ddl) CreateView(ctx sessionctx.Context, s *ast.CreateViewStmt) (err err
...
-	viewInfo, cols := buildViewInfoWithTableColumns(ctx, s)
+	viewInfo, err := buildViewInfo(ctx, s)
+	if err != nil {
+		return err
+	}
 
-	colObjects := make([]interface{}, 0, len(viewInfo.Cols))
-	for _, col := range viewInfo.Cols {
-		colObjects = append(colObjects, col)
+	var cols = make([]*table.Column, len(s.Cols))
```
Comment (zz-jason) [inline ddl/ddl_api.go]:
```suggestion
	cols := make([]*table.Column, len(s.Cols))
```

**Comment by AndrewDi** [comment]
@zz-jason @XuHuaiyu Please help to fix this old view test.

**Comment by XuHuaiyu** [comment]
@AndrewDi 
Check this test case:
``` sql
CREATE TABLE t1 (id INTEGER, sex CHAR(1));
CREATE VIEW v AS SELECT id, SUM(id) OVER w, sex FROM t1 WINDOW w AS (PARTITION BY sex ORDER BY id ROWS BETWEEN 2 PRECEDING AND 1 PRECEDING);
SELECT * FROM v;
```

**Comment by zz-jason** [comment]
@AndrewDi any update?

**Comment by AndrewDi** [comment]
@XuHuaiyu @zz-jason wild test case:
```
FAIL: logical_plan_test.go:2260: testPlanSuite.TestWindowFunction

logical_plan_test.go:2563:
    c.Assert(err.Error(), Equals, tt.result, comment)
... obtained string = "[planner:1054]Unknown column 'z' in 'field list'"
... expected string = "TableReader(Table(t))->Window(sum(cast(Column#2)) over())->Sort->Projection"
... case:6 sql:select a, b as z from t order by (sum(z) over())
```

**Comment by XuHuaiyu** [comment]
@XuHuaiyu I'll take a look at this.

**Comment by XuHuaiyu** [comment]
@AndrewDi 
The fail is caused by https://github.com/pingcap/tidb/pull/11818/files#diff-b9e75a99170ea35a67074d5d3d3305adL2084

**Comment by AndrewDi** [comment]
@XuHuaiyu Here is the wild part,if I run explain manually,everything works fine, but test always fail...

**Comment by XuHuaiyu** [comment]
@AndrewDi There is still something wrong with WindowFunc.Restore, I'll check it.

**Comment by XuHuaiyu** [comment]
@AndrewDi 
The test failed because the plan is built from the [restored SQL](https://github.com/pingcap/tidb/blob/master/planner/core/logical_plan_test.go#L2561-L2570).
But the SQL is restored wrongly because that the original AST has been modified.
The original AST is modified because [this line](https://github.com/pingcap/tidb/pull/11818/files#diff-b9e75a99170ea35a67074d5d3d3305adL2085) is deleted.

**Comment by AndrewDi** [comment]
@XuHuaiyu `WindowFunc.Restore` issue has fixed, but `Prepare func` still failed, because https://github.com/pingcap/tidb/blob/a72617d33ff49fe956b04bbbe989e499db3ada3d/planner/core/common_plans.go#L242 and https://github.com/pingcap/tidb/blob/a72617d33ff49fe956b04bbbe989e499db3ada3d/executor/prepared.go#L156 will unfold the colname.

File: planner/core/logical_plan_builder.go [hunk 9]
```diff
@@ -2084,12 +2084,12 @@ func (b *PlanBuilder) buildSelect(ctx context.Context, sel *ast.SelectStmt) (p L
 		p = b.buildTableDual()
 	}
 
-	originalFields := sel.Fields.Fields
 	sel.Fields.Fields, err = b.unfoldWildStar(p, sel.Fields.Fields)
 	if err != nil {
 		return nil, err
 	}
 
+	originalFields := sel.Fields.Fields
```
Comment (XuHuaiyu) [inline planner/core/logical_plan_builder.go]: Why not put this in line 2087?

Comment (AndrewDi) [reply planner/core/logical_plan_builder.go]: because we still need to unfold wildstart...

File: planner/core/logical_plan_builder.go [hunk 10]
```diff
@@ -2114,6 +2114,9 @@ func (b *PlanBuilder) buildSelect(ctx context.Context, sel *ast.SelectStmt) (p L
 	if err != nil {
 		return nil, err
 	}
+	if b.capFlag&canExpandAST != 0 {
+		originalFields = sel.Fields.Fields
+	}
```
Comment (XuHuaiyu) [inline planner/core/logical_plan_builder.go]: PTAL @zz-jason @winoros

Comment (eurekaka) [reply planner/core/logical_plan_builder.go]: Can we add this check for line https://github.com/pingcap/tidb/pull/11818/files#diff-b9e75a99170ea35a67074d5d3d3305adR2241 instead?

Comment (XuHuaiyu) [reply planner/core/logical_plan_builder.go]: It seems that we do not need to check it. We only need to set the `originalFields` considering capFlag.

**Pull request merged**